### PR TITLE
V1.0

### PR DESCRIPTION
**Pull request by kev-zhao**

### CODE REVIEW

**Comment by jacob-zietek** [comment]
👀

File: data.py [hunk 1]
```diff
@@ -65,3 +98,67 @@ def __getitem__(self, idx):
...
+        if c > 1:
+            print(f"Generated mask {c} > 1 times.")
+
+        x_THW = unfactorize_token_ids(x_THWC, config.num_factored_vocabs, config.factored_vocab_size)
+        x_THW[:, first_masked_frame:][mask] = mask_token_id
+
+        return {
+            "input_ids": rearrange(x_THW, "b t h w -> b (t h w)"),
+            "labels": rearrange(labels, "b t h w -> b (t h w)"),
+            # "relevant_mask": mask | random_patches_mask  # only 15 frames instead of 16 though
```
Comment (Guitaricet) [inline data.py:161]: delete

File: eval_utils.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+           f"Shape of `logits` should be (B, {factored_vocab_size}, {num_factored_vocabs}, T-1, H, W)"
+    t = factored_logits.size(3) + 1
+    h, w = factored_logits.size()[-2:]
+    assert t * h * w == labels_flat.size(1), "Shape of `factored_logits` does not match flattened latent image size."
+
+    labels_THW = rearrange(labels_flat, "b (t h w) -> b t h w", t=t, h=h, w=w)
+    labels_THW = labels_THW[:, 1:].to(factored_logits.device)
+
+    factored_labels = factorize_labels(labels_THW, num_factored_vocabs, factored_vocab_size)
+    return torch.nn.functional.cross_entropy(factored_logits, factored_labels, reduction="none").sum(dim=1).mean().item()
```
Comment (Guitaricet) [inline eval_utils.py:76]: explain reduction=None, sum=1 with a comment

File: genie/attention.py [hunk 3]
```diff
@@ -1,52 +1,44 @@
...
+    def forward(self, x: torch.Tensor, causal: bool = False) -> torch.Tensor:
         B, N, C = x.shape
-        qkv = self.qkv(x).reshape(B, N, 3, self.num_heads, C // self.num_heads)
+        qkv = self.qkv(x).reshape(B, N, 3, self.config.num_heads, self.head_dim)
         q, k, v = unbind(qkv, 2)
-        if self.qkv_norm:
+        if self.qk_norm:
             q = self.norm(q)
             k = self.norm(k)
             # LN done in float32, cast back to bf16
```
Comment (Guitaricet) [inline genie/attention.py:34]: this comment is not true, need to delete and remove explicit cast below. Or to decorate it with fp32-forcinng autocast

File: genie/st_transformer.py [hunk 4]
```diff
@@ -28,28 +23,16 @@ def forward(self, x: Tensor) -> Tensor:
...
         super().__init__()
-        self.norm1 = nn.Identity() if qkv_norm else nn.LayerNorm(dim, eps=1e-05)
-        # sequence dim is over each frame's 32x32 patch tokens
-        self.spatial_attn = SelfAttention(dim, num_heads=num_heads, qkv_bias=qkv_bias, proj_bias=proj_bias, attn_drop=attn_drop, qkv_norm=qkv_norm)
+        self.norm1 = nn.Identity() if config.qk_norm else nn.LayerNorm(config.d_model, eps=1e-05)
+        # sequence dim is over each frame's 20x20 patch tokens
+        self.spatial_attn = SelfAttention(config)
         # sequence dim is over time sequence (16)
-        self.temporal_attn = SelfAttention(dim, num_heads=num_heads, qkv_bias=qkv_bias, proj_bias=proj_bias, attn_drop=attn_drop, qkv_norm=qkv_norm)
+        self.temporal_attn = SelfAttention(config)
```
Comment (Guitaricet) [inline genie/st_transformer.py:32]: let's have a default constructor like we had before and if you want to use config, add classmethod `from_config`

File: magvit2/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# Open-MAGVIT2
+
+Code taken from [Open-MAGVIT2](https://github.com/TencentARC/Open-MAGVIT2)
```
Comment (Guitaricet) [inline magvit2/README.md:3]: add their LICENCE file to this directory to comply with Apache licence

File: genie/st_mask_git.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+    @torch.no_grad()
+    def maskgit_generate(
+        self,
+        prompt_THW,
+        out_t,
+        maskgit_steps=1,
+        temperature=0,
+    ):
+        """
+        If `temperature` <= 1e-8, will do greedy sampling.
```
Comment (Guitaricet) [inline genie/st_mask_git.py:99]: add better docstring that explains what all parameters mean

File: genie/st_mask_git.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+
+    def set_mup_shapes(self, rescale_params=False):
+        base_config = self.config.shallow_copy()
+        base_config.num_heads = 8
+        base_config.d_model = 256  # currently hardcoding to this shape
+        base_model = STMaskGIT(base_config)
+
+        mup.set_base_shapes(self, base_model, rescale_params=rescale_params)
+
+    @classmethod
```
Comment (Guitaricet) [inline genie/st_mask_git.py:245]: maybe add `@override`? `from typing import override`

Comment (kev-zhao) [reply genie/st_mask_git.py:245]: Not in Python 3.10 I think

**Pull request merged**